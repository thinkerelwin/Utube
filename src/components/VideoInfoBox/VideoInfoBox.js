import React, { useState } from 'react';
import './VideoInfoBox.scss';
import PropTypes from 'prop-types';
import { Image, Button, Divider } from 'semantic-ui-react';
import Linkify from 'react-linkify';
import { getPublishedAtDateString } from '../../services/date/date-format';
import { getShortNumberString } from '../../services/number/number-format';

export function VideoInfoBox(props) {
  const [collapsed, setCollapsed] = useState(true);

  if (!props.video || !props.channel) {
    return <div />;
  }

  const descriptionParagraphs = getDescriptionParagraphs();
  const { descriptionTextClass, buttonTitle } = getConfig();
  const publishedAtString = getPublishedAtDateString(
    props.video.snippet.publishedAt,
  );

  const onToggleCollapseButtonClick = () => {
    setCollapsed(!collapsed);
  };

  function getDescriptionParagraphs() {
    const videoDescription = props.video.snippet
      ? props.video.snippet.description
      : null;
    if (!videoDescription) {
      return null;
    }
    return videoDescription.split('\n').map((paragraph, index) => (
      <p key={index}>
        <Linkify>{paragraph}</Linkify>
      </p>
    ));
  }

  function getConfig() {
    const descriptionTextClass = collapsed ? 'collapsed' : 'expanded';
    let buttonTitle = collapsed ? 'Show More' : 'Show Less';
    return {
      descriptionTextClass,
      buttonTitle,
    };
  }

  function getSubscriberButtonText() {
    const parsedSubscriberCount = Number(
      props.channel.statistics.subscriberCount,
    );
    const subscriberCount = getShortNumberString(parsedSubscriberCount);
    return `Subscribe ${subscriberCount}`;
  }

  const buttonText = getSubscriberButtonText();
  const channelThumbnail = props.channel.snippet.thumbnails.medium.url;
  const channelTitle = props.channel.snippet.title;

  return (
    <>
      <div className="video-info-box">
        <Image className="channel-image" src={channelThumbnail} circular />
        <div className="video-info">
          <div className="channel-name">{channelTitle}</div>
          <div className="video-publication-date">{publishedAtString}</div>
        </div>
        <Button color="youtube">{buttonText}</Button>
        <div className="video-description">
          <div className={descriptionTextClass}>{descriptionParagraphs}</div>
          <Button
            className="description-toggle"
            compact
            onClick={onToggleCollapseButtonClick}
          >
            {buttonTitle}
          </Button>
        </div>
      </div>
      <Divider />
    </>
  );
}

VideoInfoBox.propTypes = {
  video: PropTypes.object,
  channel: PropTypes.object,
};
