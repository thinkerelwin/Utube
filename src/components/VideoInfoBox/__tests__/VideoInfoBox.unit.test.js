import { VideoInfoBox } from '../VideoInfoBox';
import { shallow, mount } from 'enzyme';
import React from 'react';

const mockData = {
  video: {
    id: 'TitZV6k8zfA',
    snippet: {
      publishedAt: '2024-04-14T22:41:37Z',
      channelId: 'UCBJycsmduvYEL83R_U4JriQ',
      title: "The Worst Product I've Ever Reviewed... For Now",
      description:
        "The Humane AI pin is... bad. Almost no one should buy it. Yet.\n\nMKBHD Merch: http://shop.MKBHD.com\n\nTech I'm using right now: https://www.amazon.com/shop/MKBHD\n\nIntro music: http://youtube.com/20syl\nPlaylist of MKBHD Intro music: https://goo.gl/B3AWV5\n\nPin provided by pin for review.\n\n~\nhttp://twitter.com/MKBHD\nhttp://instagram.com/MKBHD\nhttp://facebook.com/MKBHD\n\n0:00 What Is it?\n2:30 The Hardware\n4:26 What Does It Do?\n8:14 The Review (It's Bad)\n15:00 Smartphones are OP\n21:20 A Victim of Its Future Ambition",
      thumbnails: {
        medium: {
          url: 'https://i.ytimg.com/vi/TitZV6k8zfA/mqdefault.jpg',
          width: 320,
          height: 180,
        },
      },
      channelTitle: 'Marques Brownlee',
    },
    contentDetails: {
      duration: 'PT25M4S',
    },
    statistics: {
      viewCount: '3048797',
      likeCount: '114287',
      favoriteCount: '0',
      commentCount: '10146',
    },
  },
  channel: {
    id: 'UCBJycsmduvYEL83R_U4JriQ',
    snippet: {
      title: 'Marques Brownlee',
      description:
        'MKBHD: Quality Tech Videos | YouTuber | Geek | Consumer Electronics | Tech Head | Internet Personality!\n\nbusiness@MKBHD.com\n\nNYC',
      thumbnails: {
        medium: {
          url: 'https://yt3.ggpht.com/lkH37D712tiyphnu0Id0D5MwwQ7IRuwgQLVD05iMXlDWO-kDHut3uI4MgIEAQ9StK0qOST7fiA=s240-c-k-c0x00ffffff-no-rj',
          width: 240,
          height: 240,
        },
      },
    },
    statistics: {
      subscriberCount: '18600000',
    },
  },
};

describe('VideoInfoBox', () => {
  test('renders collapsed', () => {
    const wrapper = shallow(<VideoInfoBox />);
    expect(wrapper).toMatchSnapshot();
  });
  test('renders expanded', () => {
    const wrapper = mount(<VideoInfoBox {...mockData} />);
    wrapper.find('button.description-toggle').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
