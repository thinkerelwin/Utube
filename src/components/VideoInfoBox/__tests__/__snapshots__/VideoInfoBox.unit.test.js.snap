// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoInfoBox renders collapsed 1`] = `<div />`;

exports[`VideoInfoBox renders expanded 1`] = `
Array [
  <div
    className="video-info-box"
  >
    <img
      className="ui circular image channel-image"
      src="https://yt3.ggpht.com/lkH37D712tiyphnu0Id0D5MwwQ7IRuwgQLVD05iMXlDWO-kDHut3uI4MgIEAQ9StK0qOST7fiA=s240-c-k-c0x00ffffff-no-rj"
    />
    <div
      className="video-info"
    >
      <div
        className="channel-name"
      >
        Marques Brownlee
      </div>
      <div
        className="video-publication-date"
      >
        Mon Apr 15 2024
      </div>
    </div>
    <button
      className="ui youtube button"
      onClick={[Function]}
    >
      Subscribe 19M
    </button>
    <div
      className="video-description"
    >
      <div
        className="expanded"
      >
        <p
          key="0"
        >
          The Humane AI pin is... bad. Almost no one should buy it. Yet.
        </p>
        <p
          key="1"
        />
        <p
          key="2"
        >
          Array [
            "MKBHD Merch: ",
            <a
              href="http://shop.MKBHD.com"
              key="0"
            >
              http://shop.MKBHD.com
            </a>,
          ]
        </p>
        <p
          key="3"
        />
        <p
          key="4"
        >
          Array [
            "Tech I'm using right now: ",
            <a
              href="https://www.amazon.com/shop/MKBHD"
              key="0"
            >
              https://www.amazon.com/shop/MKBHD
            </a>,
          ]
        </p>
        <p
          key="5"
        />
        <p
          key="6"
        >
          Array [
            "Intro music: ",
            <a
              href="http://youtube.com/20syl"
              key="0"
            >
              http://youtube.com/20syl
            </a>,
          ]
        </p>
        <p
          key="7"
        >
          Array [
            "Playlist of MKBHD Intro music: ",
            <a
              href="https://goo.gl/B3AWV5"
              key="0"
            >
              https://goo.gl/B3AWV5
            </a>,
          ]
        </p>
        <p
          key="8"
        />
        <p
          key="9"
        >
          Pin provided by pin for review.
        </p>
        <p
          key="10"
        />
        <p
          key="11"
        >
          ~
        </p>
        <p
          key="12"
        >
          <a
            href="http://twitter.com/MKBHD"
            key="0"
          >
            http://twitter.com/MKBHD
          </a>
        </p>
        <p
          key="13"
        >
          <a
            href="http://instagram.com/MKBHD"
            key="0"
          >
            http://instagram.com/MKBHD
          </a>
        </p>
        <p
          key="14"
        >
          <a
            href="http://facebook.com/MKBHD"
            key="0"
          >
            http://facebook.com/MKBHD
          </a>
        </p>
        <p
          key="15"
        />
        <p
          key="16"
        >
          0:00 What Is it?
        </p>
        <p
          key="17"
        >
          2:30 The Hardware
        </p>
        <p
          key="18"
        >
          4:26 What Does It Do?
        </p>
        <p
          key="19"
        >
          8:14 The Review (It's Bad)
        </p>
        <p
          key="20"
        >
          15:00 Smartphones are OP
        </p>
        <p
          key="21"
        >
          21:20 A Victim of Its Future Ambition
        </p>
      </div>
      <button
        className="ui compact button description-toggle"
        onClick={[Function]}
      >
        Show Less
      </button>
    </div>
  </div>,
  <div
    className="ui divider"
  />,
]
`;
